### PR TITLE
[Truffle] Make global variables constants if they are rarely changed

### PR DESCRIPTION
**Pull request by eregon**
@jruby/truffle @thedarkone So here is my attempt to make global variables constants if they are not changed too often. The logic is unfortunately more complex than I expected, so I would appreciate a review :smiley:

### CODE REVIEW

File: core/src/main/java/org/jruby/util/cli/Options.java [hunk 1]
```diff
@@ -258,6 +258,7 @@
     public static final Option<Integer> TRUFFLE_BIND_CACHE = integer(TRUFFLE, "truffle.bind.cache", TRUFFLE_DEFAULT_CACHE.load(), "Cache size of test for being able to bind a method to a module.");
     public static final Option<Integer> TRUFFLE_CONSTANT_CACHE = integer(TRUFFLE, "truffle.constant.cache", TRUFFLE_DEFAULT_CACHE.load(), "Constant cache size.");
     public static final Option<Integer> TRUFFLE_INSTANCE_VARIABLE_CACHE = integer(TRUFFLE, "truffle.instance_variable.cache", TRUFFLE_DEFAULT_CACHE.load(), "Instance variable cache size.");
+    public static final Option<Integer> TRUFFLE_GLOBAL_VARIABLE_CACHE = integer(TRUFFLE, "truffle.global_variable.cache", 10, "Maximum number of times a global variable can be changed to be considered constant.");
```
Comment (eregon) [inline core/src/main/java/org/jruby/util/cli/Options.java]: JRuby classic uses 100 here (see above), but that seems potentially too many deopts.

File: truffle/src/main/java/org/jruby/truffle/language/globals/GlobalVariableStorage.java [hunk 2]
```diff
@@ -10,10 +10,62 @@
...
+import org.jruby.util.cli.Options;
+
 public class GlobalVariableStorage {
-    public volatile Object value;
+
+    private static final int GLOBAL_VARIABLE_CACHE = Options.TRUFFLE_GLOBAL_VARIABLE_CACHE.load();
+
+    private final CyclicAssumption unchangedAssumption = new CyclicAssumption("global variable unchanged");
+    private int changes = 0;
+    @CompilationFinal private volatile boolean assumeConstant = true;
```
Comment (eregon) [inline truffle/src/main/java/org/jruby/truffle/language/globals/GlobalVariableStorage.java]: This really means `@CompilationFinal` for compilation and `volatile` in interpreter.

File: core/src/main/java/org/jruby/util/cli/Options.java (hunk 1 above)
Comment (chrisseaton) [inline core/src/main/java/org/jruby/util/cli/Options.java]: This should perhaps be called `_INVALIDATIONS` and `.invalidations` as it isn't the size of a cache, like the other `_CACHE` variable are.

File: truffle/src/main/java/org/jruby/truffle/language/globals/GlobalVariableStorage.java (hunk 2 above)
Comment (chrisseaton) [inline truffle/src/main/java/org/jruby/truffle/language/globals/GlobalVariableStorage.java]: Maybe put that in a comment.

**Comment by thedarkone** [comment]
lgtm.

**Pull request merged**